src/jumps.jl

```diff
@@ -35,29 +35,86 @@ end
 RegularJump(rate,c,dc;mark_dist = nothing,constant_c = false) =
             RegularJump(rate,c,dc,mark_dist,constant_c)
 
-struct JumpSet{T1,T2,T3} <: AbstractJump
+
+struct MassActionJump{T,S} <: AbstractJump
+  scaled_rates::T
+  reactant_stoch::S
+  net_stoch::S
+
+  function MassActionJump{T,S}(rates::T, rs::S, ns::S, scale_rates::Bool) where {T,S}
+    sr = copy(rates)
+    if scale_rates && !isempty(sr) 
+      scalerates!(sr, rs)
+    end
+    new(sr, rs, ns)
+  end
+end
+MassActionJump(usr::T, rs::S, ns::S; scale_rates = true ) where {T,S} = MassActionJump{T,S}(usr, rs, ns, scale_rates)
+
+
+struct JumpSet{T1,T2,T3,T4} <: AbstractJump
   variable_jumps::T1
   constant_jumps::T2
   regular_jump::T3
+  massaction_jump::T4
 end
 
-JumpSet(jump::ConstantRateJump) = JumpSet((),(jump,),nothing)
-JumpSet(jump::VariableRateJump) = JumpSet((jump,),(),nothing)
-JumpSet(jump::RegularJump) = JumpSet((),(),jump)
-JumpSet() = JumpSet((),(),nothing)
+JumpSet(jump::ConstantRateJump) = JumpSet((),(jump,),nothing,nothing)
+JumpSet(jump::VariableRateJump) = JumpSet((jump,),(),nothing,nothing)
+JumpSet(jump::RegularJump)      = JumpSet((),(),jump,nothing)
+JumpSet(jump::MassActionJump)   = JumpSet((),(),nothing,jump)
+JumpSet() = JumpSet((),(),nothing,nothing)
 JumpSet(jb::Void) = JumpSet()
 
 # For Varargs, use recursion to make it type-stable
 
-JumpSet(jumps::AbstractJump...) = JumpSet(split_jumps((), (), nothing, jumps...)...)
+JumpSet(jumps::AbstractJump...) = JumpSet(split_jumps((), (), nothing, nothing, jumps...)...)
```

Comment:
I have a feeling that building this with a ton of jumps will be terrible for performance. One of the next things we should look at is allowing vectors of jumps as input.

Reply:
Yup. Not an issue for the `MassActionJump`, but it would be an issue for `ConstantJumps` or `VariableRateJumps`. It does seem that everything works if one just directly creates a `JumpSet` that stores `ConstantJumps` as a vector, and then passes that into the `JumpProblem`. I do that in the new test script I added.